core/src/main/java/org/infinispan/interceptors/InvocationStage.java

```diff
@@ -12,74 +23,248 @@
+    * {@link BaseAsyncInterceptor#invokeNext(InvocationContext, VisitableCommand)} to invoke a new command.
+    */
+   <P1> InvocationStage thenCompose(P1 p1, BiFunction<P1, Object, InvocationStage> function);
+
+   /**
+    * Execute a callback after the stage has finished with an exception.
+    * <p>
+    * The callback can rethrow the exception, throw a new one, or return a regular value.
+    */
+   <P1> InvocationStage exceptionally(P1 p1, BiFunction<P1, Throwable, Object> function);
+
+   /**
+    * Execute a callback after the stage has finished with an exception.
+    * <p>
+    * The callback can rethrow the exception, throw a new one, or return another {@link InvocationStage}, e.g.
+    * using {@link BaseAsyncInterceptor#invokeNext(InvocationContext, VisitableCommand)} to invoke a new command.
+    */
+   <P1> InvocationStage exceptionallyCompose(P1 p1, BiFunction<P1, Throwable, InvocationStage> function);
+
+   /**
+    * Execute a callback after the stage has finished (P1 p1, either successfully or with an exception).
+    * <p>
+    * The callback can not change the return value, but it can throw an exception.
+    */
+   <P1> InvocationStage handle(P1 p1, TriFunction<P1, Object, Throwable, Object> function);
+
+   /**
+    * Execute a callback after the stage has finished (P1 p1, either successfully or with an exception).
+    * <p>
+    * The callback can rethrow the exception, throw a new one, or return a regular value.
+    */
+   <P1> InvocationStage whenComplete(P1 p1, TriConsumer<P1, Object, Throwable> action);
+
+   /**
+    * Execute a callback after the stage has finished (P1 p1, either successfully or with an exception).
+    * <p>
+    * The callback can return any {@link InvocationStage}, e.g. using
+    * {@link BaseAsyncInterceptor#invokeNext(InvocationContext, VisitableCommand)} to invoke a new command.
+    */
+   <P1> InvocationStage compose(P1 p1, TriFunction<P1, Object, Throwable, InvocationStage> function);
+
+
+   // Methods with 2 external parameters
+
+   /**
+    * Execute a callback after stage has finished successfully.
+    * <p>
+    * It can change the return value, or it can throw an exception.
+    */
+   <P1, P2> InvocationStage thenApply(P1 p1, P2 p2, TriFunction<P1, P2, Object, Object> function);
+
+   /**
+    * Execute a callback after the stage has finished successfully.
+    * <p>
+    * The callback can not change the return value, but it can throw an exception.
+    */
+   <P1, P2> InvocationStage thenAccept(P1 p1, P2 p2, TriConsumer<P1, P2, Object> action);
+
+   /**
+    * Execute a callback after the stage has finished successfully.
+    * <p>
+    * The callback can return any {@link InvocationStage}, e.g. using
+    * {@link BaseAsyncInterceptor#invokeNext(InvocationContext, VisitableCommand)} to invoke a new command.
+    */
+   <P1, P2> InvocationStage thenCompose(P1 p1, P2 p2, TriFunction<P1, P2, Object, InvocationStage> function);
```

Comment:
what is the goal of all theses P1 and P2?
If it is to capture the context and command, doesn't make more sense to use InvocationContext and VisitableCommand?
If it is to capture any variable, only 2 parameters isn't enough.